Add JanusGraph specific authentication.

krlohnes opened the pull request:
Issue: #623

This addresses a few concerns.

1) Configure a credential graph that automatically creates the appropriate JanusGraph specific indexes.

2) TinkerPop now supports (as of 3.2.6 and 3.3.0) a combined channelizer
for enabling port unification for WebSockets and HTTP. This adds the
necessary classes to enable authentication for one, the other, or both
using JanusGraph specific authentication.

3) HTTP Basic Auth is (and should be) slow. This adds an HMAC token
authentication mechanism to allow for fast authentication over HTTP.

Thank you for contributing to JanusGraph.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there an issue associated with this PR? Is it referenced in the commit message?
- [X] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [X] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [X] Have you written and/or updated unit tests to verify your changes?
- [ N/A] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ N/A] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ N/A] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered? 

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

mbrukman commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/HMACAuthenticator.java:
```diff
@@ -0,0 +1,225 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package org.janusgraph.graphdb.tinkerpop.gremlin.server.auth;
```
Please add a blank line below the license, before the package name — some IDEs hide the top-level boilerplate comments, and lacking a blank line here may cause them to hide this package line as well.

mbrukman commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphAbstractAuthenticator.java:
```diff
@@ -0,0 +1,130 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ * Modifications Copyright 2017 JanusGraph Authors
```
Was this file branched from another file in this (or another) repo, and then modified? Would it be possible to create a single commit which is just the clean copy of the file (renamed, perhaps), and the next commit adds the modifications so it's clear what the changes are, or is that too much work / too complicated?

krlohnes replied:
I wasn't really sure how to handle this. It borrows heavily from the [TinkerPop SimpleAuthenticator](https://github.com/apache/tinkerpop/blob/3.2.5/gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/auth/SimpleAuthenticator.java)

Maybe it should just list the class it modified? It's also an abstracts away the `JanusGraphSimpleAuthenticator` and the `SaslAndHMACAuthenticator` which also borrow heavily from that class. Would it make sense to list he TinkerPop version and that it borrows from `org.apache.tinkerpop.gremlin.server.auth.SimpleAuthenticator.java`

sjudeng replied:
If `credentialsStore` in `SimpleAuthenticator` was protected instead of private (or equivalently if it was obtained through a protected accessor in `SimpleAuthenticator#authenticate`) would you have just been able to collapse `JanusGraphAbstractAuthenticator` and `JanusGraphSimpleAuthenticator` into one, overriding `setup()`, and then update your other implementations to further extend same?

If so then what about in `JanusGraphAbstractAuthenticator#setup` (or after collapse `JanusGraphSimpleAuthenticator#setup`) use reflection to the set `SimpleAuthenticator.credentialsStore`? Reflection isn't very clean but I'd prefer it here over all the code duplication and license header issues, both of which would go away with this update. You could then reach out to TinkerPop about updates to better support auth extensions and contribute that as a separate PR to them so eventually we could remove the need for reflection on the JanusGraph side.

krlohnes replied:
@sjudeng This would only address the JanusGraphSimpleAuthenticator, not the `JanusGraphAbstractAuthenticator`. There's also the `SaslAndHMACAuthenticator`, which provides authentication for both HTTP and WebSockets for the combined channelizer (`WsAndHttpChannelizer`). So I could change the `JanusGraphSimpleAuthenticator` to use the TinkerPop `SimpleAuthenticator` through reflection, but I wouldn't be able to _collapse_ the two classes, so it wouldn't fully alleviate the issue. 

It did get me thinking, though, that I'd seen examples of the `GraphFactory` opening JanusGraph properties files. I tested that locally, and it seems to work. We could just open the graph, create the indices, close the graph instance and clear the reference and let the SimpleAuthenticator take care of everything after `super.setup()` is called. I'll give it whirl and check back in with the results.

As far as the license at the top... it is copying a lot, but it's also implementing an interface from TinkerPop. Thinking about it now, I think it would be fine to just include the normal JanusGraph header license. It's implementing the API interface, it's bound to be extremely similar in the way it works. @sjudeng @mbrukman What do you think about that?

mbrukman commented on janusgraph-test/src/test/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/handler/HttpHMACAuthenticationHandlerTest.java:
```diff
@@ -0,0 +1,198 @@
+
+        try {
+            expect(authenticator.authenticate(and(isA(Map.class), capture(credMap)))).andReturn(new AuthenticatedUser("foo"));
+        } catch (Exception e) {}
```
This is masking all sorts of possible exceptions, quietly.

Why not just annotate the method as being allowed to throw exceptions (but it shouldn't in practice, right?), and drop the try/catch clause? Similarly elsewhere.

krlohnes replied:
Absolutely. I meant to change this before I committed, but it's been sitting around for a while and I forgot.

mbrukman commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/SaslAndHMACAuthenticator.java:
```diff
@@ -0,0 +1,75 @@
+ *
+ * Modifications Copyright 2017 JanusGraph Authors
+ */
+package org.janusgraph.graphdb.tinkerpop.gremlin.server.auth;
```
Blank line before package declaration, please.

pluradj commented:
@krlohnes you're probably already aware that we're in the process of getting 0.2.0 release completed, so if this PR is accepted, it won't be merged until after the release is complete. Thanks.

krlohnes commented:
@pluradj Thanks, I've been following along on the dev list. I just finally had a few minutes to push this PR and wanted to take advantage of it.

krlohnes commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/SaslAndHMACAuthenticator.java:
```diff
@@ -0,0 +1,76 @@
+ * and an HMACAuthenticator (for http)
+ * @author Keith Lohnes lohnesk@gmail.com
+ */
+public class SaslAndHMACAuthenticator extends JanusGraphAbstractAuthenticator {
```
@sjudeng Nice catch, this should be the normal JanusGraph license header.

krlohnes replied:
Fixed

robertdale commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphAbstractAuthenticator.java:
```diff
@@ -0,0 +1,130 @@
+                try {
+                    mgmt = (ManagementSystem) graph.openManagement();
+                    mgmt.updateIndex(mgmt.getGraphIndex("byUsername"), SchemaAction.REINDEX);
+                    mgmt.awaitGraphIndexStatus(graph, "byUsername").status(SchemaStatus.ENABLED).call();
```
This is a static call. Use `ManagementSystem.await...`

robertdale commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphAbstractAuthenticator.java:
```diff
@@ -0,0 +1,130 @@
+            }
+        }
+
+        if (credentialStore.countUsers() == 0) {
```
This calls `hasLabel()` and is a potential huge performance hit when there are users populated. I'm not sure I agree with creating a default user.  But you could move this up as a part of creating the schema.

krlohnes replied:
There's a unique constraint on the index, I could try catch on the user creation instead of this check

robertdale replied:
No, creating the default user is only good when creating the initial schema.  The default user should/would be removed when going to a deployed system.  You don't want to end up unwittingly recreate the default user every restart thus creating a security hole.  Maybe even attempting to create the default user should be a configuration - optional and false by default.

robertdale commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphSimpleAuthenticator.java:
```diff
@@ -0,0 +1,151 @@
+import org.apache.tinkerpop.gremlin.server.auth.AuthenticatedUser;
+import org.apache.tinkerpop.gremlin.structure.Graph;
+import org.apache.tinkerpop.gremlin.structure.Vertex;
+import org.apache.tinkerpop.gremlin.structure.io.IoCore;
```
Clean up all unused imports.

robertdale commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphSimpleAuthenticator.java:
```diff
@@ -0,0 +1,151 @@
+
+package org.janusgraph.graphdb.tinkerpop.gremlin.server.auth;
+
+import org.apache.tinkerpop.gremlin.groovy.plugin.dsl.credential.CredentialGraph;
```
Lots of deprecated. Use current replacements.

robertdale commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/SaslAndHMACAuthenticator.java:
```diff
@@ -0,0 +1,76 @@
+
+import org.apache.tinkerpop.gremlin.server.auth.AuthenticatedUser;
+import org.apache.tinkerpop.gremlin.server.auth.AuthenticationException;
+import org.slf4j.Logger;
```
Remove unused imports.

robertdale commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/SaslAndHMACAuthenticator.java:
```diff
@@ -0,0 +1,76 @@
+        "This exception is likely due to a misconfiguration. Try using the HMACAndSaslAuthenticationHandler as the " +
+        "authenticationHandler in the server authentication configuration";
+
+    private Map<String, String> authenticatorConfig;
```
Remove unused field

robertdale commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/handler/HttpHMACAuthenticationHandler.java:
```diff
@@ -0,0 +1,153 @@
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
```
Clean up lots of unused imports

robertdale commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/handler/HttpHMACAuthenticationHandler.java:
```diff
@@ -0,0 +1,153 @@
+import static io.netty.handler.codec.http.HttpResponseStatus.UNAUTHORIZED;
+import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+
+import static org.apache.tinkerpop.gremlin.groovy.plugin.dsl.credential.CredentialGraphTokens.PROPERTY_PASSWORD;
```
Replace deprecated with current

robertdale commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/handler/SaslAndHMACAuthenticationHandler.java:
```diff
@@ -0,0 +1,66 @@
+import org.apache.tinkerpop.gremlin.server.auth.Authenticator;
+import org.apache.tinkerpop.gremlin.server.handler.SaslAuthenticationHandler;
+import org.janusgraph.graphdb.tinkerpop.gremlin.server.auth.SaslAndHMACAuthenticator;
+import org.janusgraph.graphdb.tinkerpop.gremlin.server.auth.JanusGraphSimpleAuthenticator;
```
Clean up unused imports

krlohnes commented:
@robertdale 

>1. TinkerPop authentication can't be set up to use a JanusGraph graph
for storing credentials. This fixes that.
Why do you make this claim when it can be used? Can you clarify what you mean?

You're right, this isn't correct. At this point, I'd say instead. "Configure a credential graph that automatically creates the appropriate JanusGraph specific indexes". I'll fix this with when I get to the code changes.

>Can you clarify this statement? I thought you stated in PR apache/tinkerpop#618 to TinkerPop that authentication already works for both. I wonder if we need to go back and fix the documentation.

Yes, sorry if this wasn't clear. You're absolutely right, those work for TinkerPop with the current auth. This is adding support for both for the JanusGraph specific authenticators. 

>It would be interesting to see this difference quantified. Have you tried using fewer rounds when creating a user? BCrypt.hashpw(password, BCrypt.gensalt(BCRYPT_ROUNDS))

I would like to use _more_ rounds. I think lower the BCrypt rounds, from a security standpoint, isn't the right way to think about this, it changes with hardware. It could work for a single sign in session on webserver, but that's not what's happening with the Gremlin server. Really, you'd want to target an amount of time it would take for BCrypt to hash the password, so the BCrypt rounds changes for hardware. Making the BCrypt rounds configurable is something I'd like to do in TinkerPop, but haven't had time to make that change yet. 

Now, say I want to target 240ms for BCrypt to do its thing. That's an additional 240 ms of overhead that's going to happen per gremlin query, which may or may not be acceptable. HMAC allows you amortize that per request overhead in a secure fashion.

>Why is there so much duplication of code when you had created an 'abstraction to easily allow different HttpAuth schemes' in PR apache/tinkerpop#583 ?
I'm wondering if some code is better served by being in TinkerPop

I addressed some of this in a comment to sjudeng [here](https://github.com/JanusGraph/janusgraph/pull/624/files#r145737755) He made a good point, and this is also when I realized you _could_ use a JanusGraph graph with TinkerPop auth. I wanted to get his feedback on that implementation before I made code changes. There's a lot of code deduplication that could happen through that. This code change wouldn't have worked without the abstraction split between Authenticator and AuthenticationHandler apache/tinkerpop#583.

re: Documentation

In my initial message regarding documentation, I asked for feedback on where it should go (directory, should it go somewhere that already exists? New file? Should it be a subsection? Does it deserve it's own section?). Do you have feedback? I definitely want to make sure everything around this feature set is clear.

I'm traveling right now, but I'll get to the rest of your feedback on code changes (unused imports, final, static methods etc) as soon as I can. Thanks for your review.

krlohnes commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/HMACAuthenticator.java:
```diff
@@ -0,0 +1,226 @@
+ *
+ * @author Keith Lohnes lohnesk@gmail.com
+ */
+public class HMACAuthenticator extends JanusGraphAbstractAuthenticator {
```
Fixed

krlohnes commented on janusgraph-core/pom.xml:
```diff
@@ -34,6 +34,11 @@
         </dependency>
         <dependency>
             <groupId>org.apache.tinkerpop</groupId>
+            <artifactId>gremlin-server</artifactId>
```
I'm hopefully going to creating `janusgraph-server` tomorrow.

krlohnes commented:
@sjudeng @mbrukman @robertdale I've made the changes that you've suggested and added some documentation.

krlohnes commented on janusgraph-server/pom.xml:
```diff
@@ -0,0 +1,47 @@
+  <parent>
+    <groupId>org.janusgraph</groupId>
+    <artifactId>janusgraph</artifactId>
+    <version>0.2.0-SNAPSHOT</version>
```
Fixed

krlohnes commented on janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphSimpleAuthenticator.java:
```diff
@@ -0,0 +1,64 @@
+     * Created for testing
+     * @return
+     */
+    public SimpleAuthenticator createSimpleAuthenticator() {
```
@sjudeng In this case, it's used in the unit tests for creating a partial mock. We don't actually want to have to create an instance of the SimpleAuthenticator in the unit tests since it's outside the scope of the unit being tested. I've seen this type of pattern before, though I'm forgetting now if it's in TinkerPop or JanusGraph.

sjudeng replied:
I see now, thanks. Can you change to `protected` and also remove (or rephrase) the testing comment since it's not just used for testing?

krlohnes commented on janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/SaslAndHMACAuthenticator.java:
```diff
@@ -0,0 +1,75 @@
+ * and an HMACAuthenticator (for http)
+ * @author Keith Lohnes lohnesk@gmail.com
+ */
+public class SaslAndHMACAuthenticator extends JanusGraphAbstractAuthenticator {
```
Sorry, I thought I'd updated this header. I'll fix that.

krlohnes replied:
fixed

krlohnes commented on janusgraph-server/pom.xml:
```diff
@@ -0,0 +1,47 @@
+      <dependency>
+          <groupId>org.easymock</groupId>
+          <artifactId>easymock</artifactId>
+          <version>3.4</version>
```
fixed

krlohnes commented on janusgraph-server/pom.xml:
```diff
@@ -0,0 +1,47 @@
+      </dependency>
+      <dependency>
+          <groupId>org.mockito</groupId>
+          <artifactId>mockito-all</artifactId>
```
fixed

krlohnes commented on janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphAbstractAuthenticator.java:
```diff
@@ -0,0 +1,121 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+abstract public class JanusGraphAbstractAuthenticator implements Authenticator {
```
fixed

amcp replied:
add a class-level javadoc?

krlohnes commented:
@sjudeng I went through and fixed or addressed the issues you pointed out.

sjudeng commented:
Can you update .travis.yml to run the janusgraph-server tests? Also when I try to run tests locally I get an error. It looks like this can be resolved by adding janusgraph-test as a test-scoped dependency.

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.15:test (default-test) on project janusgraph-server: ExecutionException; nested exception is java.util.concurrent.ExecutionException: java.lang.RuntimeException: There was an error in the forked process
[ERROR] java.lang.RuntimeException: Unable to load category: org.janusgraph.testcategory.PerformanceTests
```

I think you also need to add log4j.properties as a test resource. There are a bunch of error messages like the following.

```
log4j:ERROR Could not read configuration file from URL [file:/home/test/janusgraph/janusgraph-server/target/test-classes/log4j.properties].
java.io.FileNotFoundException: /home/test/janusgraph/janusgraph-server/target/test-classes/log4j.properties (No such file or directory)
```

sjudeng commented on janusgraph-server/pom.xml:
```diff
@@ -0,0 +1,49 @@
+<?xml version="1.0"?>
+<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <modelVersion>4.0.0</modelVersion>
```
Use indent size of 4 spaces.

sjudeng commented on janusgraph-server/src/test/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/SaslAndHMACAuthenticatorTest.java:
```diff
@@ -0,0 +1,41 @@
+import org.junit.Test;
+
+public class SaslAndHMACAuthenticatorTest extends EasyMockSupport {
+
```
Can you add a test to cover the setup method?

sjudeng commented on docs/basics.txt:
````diff
@@ -635,23 +635,140 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
 
+```
+curl -v http://localhost:8182/session -XPOST -d '{"gremlin": "1+1"}' -H "Authorization: Token dXNlcjoxNTA5NTQ2NjI0NDUzOkhrclhYaGhRVG9KTnVSRXJ5U2VpdndhalJRcVBtWEpSMzh5WldqRTM4MW89"
+```
````
I was still able to authenticate with just username password with this configuration. Also it works with the full token but I found it also works when I change some of the trailing characters in the token. Is this expected?

robertdale replied:
That doesn't sound right at all. That's a serious flaw and needs to be resolved.

krlohnes replied:
@robertdale You're right, that isn't correct. I assumed and well, we all know what happens when you assume. My apologies to you and @sjudeng . 

I was looking in to this, and this _isn't_ the time. This is the [base64 output padding](https://en.wikipedia.org/wiki/Base64#Output_padding).  

Note if you delete the last trailing `=` it doesn't work correctly. If you delete both, it does. You can also change the 5 last trailing characters (aside from the `==`), and things will work. This _isn't_ changing any of the significant parts of the HMAC token.

krlohnes replied:
@robertdale I just saw your other comment re: HMAC as well, I'll fix that today.

krlohnes commented on docs/basics.txt:
````diff
@@ -635,23 +635,140 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
-Make sure the port parameter in each yaml configuration is different if starting more than one JanusGraph Server on the same machine:
+```
+ graphs: {
+     credentialGraph: conf/janusgraph-cassandra.properties
````
>"graph" instead of "credentialGraph"

This was to specifically show that you can use a separate graph for credentials. I could change it to 

```
graph: conf/janusgraph-cassandra.properties,
credentialGraph: conf/some-other-conf.properties
```
>g.V().count() instead of 2+2/1+1

Not everyone has a `g` variable configured. If either `2+2` or `1+1` are evaluated, you've successfully authenticated.

I'll fix the properties file to hbase properties file.

sjudeng replied:
This point in the docs follows the sections where the server with `g` is defined. I think it makes sense for this section to build on that to demonstrate adding authentication and then re-executing the graph query. But I'm more interested in there being a graph query rather than `2+2` than in whether the variable is `g` or something else (to demonstrate the difference) if that makes sense.

krlohnes commented on docs/basics.txt:
```diff
@@ -635,23 +635,140 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+   config: {
+     defaultUsername: user,
+     defaultPassword: password,
+     credentialsDb: conf/janusgraph-cassandra.properties
```
Yes, it does so you can reference that and add users, change passwords, etc. I suppose you could always open it up. I'll add some additional documentation.

sjudeng commented:
When I build the release distribution (`mvn clean install -DskipTests -Dgpg.skip=true -Pjanusgraph-release`) the janusgraph-server jar is missing. It can be resolved by adding the dependency to`janusgraph-all/pom.xml`. 

```xml
<dependency>
  <groupId>org.janusgraph</groupId>
  <artifactId>janusgraph-server</artifactId>
  <version>${project.version}</version>
</dependency>
```

robertdale commented:
I just tried following the docs and got an error.
With this config:
```
graphs: {
  graph: conf/gremlin-server/janusgraph-cassandra-es-server.properties,
  credentialGraph: conf/janusgraph-cassandra.properties }
authentication: {
  authenticator: org.janusgraph.graphdb.tinkerpop.gremlin.server.auth.HMACAuthenticator,
  authenticationHandler: org.janusgraph.graphdb.tinkerpop.gremlin.server.handler.HttpHMACAuthenticationHandler,
  config: {
    defaultUsername: user,
    defaultPassword: password,
    hmacSecret: secret,
    credentialsDb: conf/janusgraph-cassandra.properties } }
```
**Executed** `curl -v -XPOST http://localhost:8182 -d '{"gremlin": "2+2"}'`
**Expected** `HTTP 401`
**Got**:
```
Note: Unnecessary use of -X or --request, POST is already inferred.
* Rebuilt URL to: http://localhost:8182/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8182 (#0)
> POST / HTTP/1.1
> Host: localhost:8182
> User-Agent: curl/7.53.1
> Accept: */*
> Content-Length: 18
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 18 out of 18 bytes
* Recv failure: Connection reset by peer
* stopped the pause stream!
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
```
**gremlin-server.log:**
```
148319 [gremlin-server-worker-1] WARN  io.netty.channel.ChannelInitializer  - Failed to initialize a channel. Closing: [id: 0x68df4a04, L:/0:0:0:0:0:0:0:1:8182 - R:/0:0:0:0:0:0:0:1:53370]
io.netty.channel.ChannelPipelineException: org.janusgraph.graphdb.tinkerpop.gremlin.server.handler.HttpHMACAuthenticationHandler is not a @Sharable handler, so can't be added or removed multiple times.
	at io.netty.channel.DefaultChannelPipeline.checkMultiplicity(DefaultChannelPipeline.java:581)
	at io.netty.channel.DefaultChannelPipeline.addLast(DefaultChannelPipeline.java:196)
	at io.netty.channel.DefaultChannelPipeline.addLast(DefaultChannelPipeline.java:189)
	at org.apache.tinkerpop.gremlin.server.channel.WebSocketChannelizer.configure(WebSocketChannelizer.java:113)
	at org.apache.tinkerpop.gremlin.server.AbstractChannelizer.initChannel(AbstractChannelizer.java:155)
	at org.apache.tinkerpop.gremlin.server.AbstractChannelizer.initChannel(AbstractChannelizer.java:69)
	at io.netty.channel.ChannelInitializer.initChannel(ChannelInitializer.java:113)
	at io.netty.channel.ChannelInitializer.handlerAdded(ChannelInitializer.java:105)
	at io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:590)
	at io.netty.channel.DefaultChannelPipeline.access$000(DefaultChannelPipeline.java:44)
	at io.netty.channel.DefaultChannelPipeline$PendingHandlerAddedTask.execute(DefaultChannelPipeline.java:1355)
	at io.netty.channel.DefaultChannelPipeline.callHandlerAddedForAllHandlers(DefaultChannelPipeline.java:1090)
	at io.netty.channel.DefaultChannelPipeline.invokeHandlerAddedIfNeeded(DefaultChannelPipeline.java:640)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:452)
	at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:374)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:424)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:399)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:464)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:131)
	at java.lang.Thread.run(Thread.java:748)
```

robertdale commented:
Same error as above with config:
```
  authenticator: org.janusgraph.graphdb.tinkerpop.gremlin.server.auth.JanusGraphSimpleAuthenticator,
  authenticationHandler: org.apache.tinkerpop.gremlin.server.handler.HttpBasicAuthenticationHandler,
```

robertdale commented:
I deleted my two previous comments due to user error and not using the HttpChannelizer. :smiley_cat:

robertdale commented on janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/HMACAuthenticator.java:
```diff
@@ -0,0 +1,222 @@
+        } else {
+            //Don't short circuit comparison to prevent timing attacks
+            boolean isValid = true;
+            for (int i = 0; i < token.length(); i++) {
```
Here the HMAC is rendered useless because it is short circuited when the incoming length truncates it. 
You need to check first that `expected.length() == base64Token.length()`
I would drive the loop by `expected.length()`.
Either way, it would seem that you should use either `token` or `base64Token` but not both.

robertdale commented on janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/HMACAuthenticator.java:
```diff
@@ -0,0 +1,222 @@
+            boolean isValid = true;
+            for (int i = 0; i < token.length(); i++) {
+                if (base64Token.charAt(i) != expected.charAt(i)) {
+                    isValid = false;
```
You can `break` out of the loop here.

krlohnes replied:
@robertdale You can't, that opens up the algorithm to timing attacks.

krlohnes commented:
@sjudeng I addressed the testing and documentation concerns you had. I've also added the server module for travis, the server dependency in the janusgraph-all pom.xml and the log4j.properties in the test resources for the server module. I'll keep an eye on the travis tests and make sure things work out there. @robertdale I've addressed /fixed your comments on the HMAC token and validation. Thank you both for your reviews.

sjudeng commented on docs/basics.txt:
````diff
@@ -635,23 +635,144 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+should return a 401 if the authentication is configured correctly and
+
+```
+curl -v -XPOST http://localhost:8182 -d '{"gremlin": "g.V().count()"}' -u user:password
````
Here I get an error (`No such property: g for class: Script3`). Is there a missing step to get graph queries to work? I can resolve by changing `credentialsGraph` to `graph` but is there another way since you want to keep the credentialsGraph name?

krlohnes replied:
@sjudeng I've removed credential graph and added an `IMPORTANT` block in each section stating that the `credentialsDb` should be different from the users `graphs`.

sjudeng replied:
Thanks. To avoid confusion can you now also change `conf/gremlin-server/http-janusgraph-hbase-server.properties` here and below to something else to make it clear it should be different? Maybe `conf/janusgraph-credentials-server.properties` or something. Also add documentation indicating that the properties file should define appropriate storage backend (and index backend if necessary) and stating what it will be used for (e.g. store credentials information).

krlohnes replied:
Fixed, assuming the wording is okay for you.

sjudeng commented on docs/basics.txt:
````diff
@@ -635,23 +635,144 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+ }
+```
+
+Note that credentialGraph and credentialsDb should point to the same properties file. `credentialGraph` is not required, but make it easier to have the graph available to make changes through the http://tinkerpop.apache.org/docs/3.2.6/reference/#credentials-dsl[TinkerPop CredentialGraph DSL]
````
It looks like the implementation creates user nodes in the graph? Then shouldn't `credentialsDb` and `credentialGraph` be different? Otherwise user nodes are added to the production graph. I would think many users wouldn't expect this, just from a data modeling point of view. Also this exposes usernames and (albeit encrypted) passwords through the server endpoint.

sjudeng commented on docs/basics.txt:
````diff
@@ -635,23 +635,144 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+
+```
+graphs: {
+    credentialGraph: conf/gremlin-server/http-janusgraph-hbase-server.properties
````
/http/socket/

sjudeng commented on docs/basics.txt:
```diff
@@ -635,23 +635,144 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+  config: {
+    defaultUsername: user,
+    defaultPassword: password,
+    credentialsDb: conf/gremlin-server/http-janusgraph-hbase-server.properties
```
/http/socket/

robertdale commented on docs/basics.txt:
````diff
@@ -635,23 +635,144 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+curl -v http://localhost:8182/session -XPOST -d '{"gremlin": "1+1"}' -H "Authorization: Token dXNlcjoxNTA5NTQ2NjI0NDUzOkhrclhYaGhRVG9KTnVSRXJ5U2VpdndhalJRcVBtWEpSMzh5WldqRTM4MW89"
+```
+
+==== Authentication over WebSockets
````
WebSocket is a proper noun, do not pluralise everythings.

robertdale commented on docs/basics.txt:
````diff
@@ -635,23 +635,144 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+```
+
+==== Authentication over WebSockets
+Authentication over WebSockets occurs through a Simple Authentication and Security Layer (https://en.wikipedia.org/wiki/Simple_Authentication_and_Security_Layer[SASL]) mechanism.
````
WebSocket

robertdale commented on docs/basics.txt:
````diff
@@ -635,23 +635,144 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+password: password
+```
+
+==== Authentication over HTTP and WebSockets
````
WebSocket

robertdale commented on docs/basics.txt:
````diff
@@ -635,23 +635,144 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+```
+
+==== Authentication over HTTP and WebSockets
+If you are using the combined channelizer for both HTTP and WebSockets you can use the SaslAndHMACAuthenticator to authorize through either WebSocket through SASL, HTTP through basic auth, and HTTP through HMAC token described previously.
````
WebSocket

krlohnes commented:
@robertdale made the WebSockets -> WebSocket change

sjudeng commented on docs/basics.txt:
````diff
@@ -635,22 +635,131 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+You can then use that token for authentication by using the "Authorization: Token" header. For example
+
+```
+curl -v http://localhost:8182/session -XPOST -d '{"gremlin": "1+1"}' -H "Authorization: Token dXNlcjoxNTA5NTQ2NjI0NDUzOkhrclhYaGhRVG9KTnVSRXJ5U2VpdndhalJRcVBtWEpSMzh5WldqRTM4MW89"
````
Use `g.V().count()` here as well

sjudeng commented on docs/basics.txt:
```diff
@@ -635,22 +635,131 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+  config: {
+    defaultUsername: user,
+    defaultPassword: password,
+     credentialsDb: conf/janusgraph-credential-server.properties
```
"janusgraph-credential**s**-server.properties" and remove leading extra space

sjudeng commented on docs/basics.txt:
````diff
@@ -635,22 +635,131 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+```
+authentication: {
+  authenticator: org.janusgraph.graphdb.tinkerpop.gremlin.server.auth.JanusGraphSimpleAuthenticator,
+  authenticationHandler: org.apache.tinkerpop.gremlin.server.handler.HttpBasicAuthenticationHandler,
````
I'm getting an error with this config when trying to connect in Gremlin console. Should it be the SASL authenticator[Handler] here?

```
io.netty.channel.ChannelPipelineException: org.apache.tinkerpop.gremlin.server.handler.HttpBasicAuthenticationHandler is not a @Sharable handler, so can't be added or removed multiple times.
```

sjudeng commented on docs/basics.txt:
```diff
@@ -635,22 +635,131 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+    defaultUsername: user,
+    defaultPassword: password,
+    hmacSecret: secret,
+    credentialsDb: conf/janusgraph-configuration-server.properties
```
"conf/janusgraph-credentials-server.properties"

krlohnes commented:
> Some comments from integration testing with WebSocket.

Fixed

sjudeng commented on docs/basics.txt:
```diff
@@ -635,22 +635,131 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+  config: {
+    defaultUsername: user,
+    defaultPassword: password,
+     credentialsDb: conf/janusgraph-credentials-server.properties
```
Remove extra leading space

krlohnes replied:
fixed

krlohnes commented:
@sjudeng I fixed that issue with the combined HTTP/WebSockets and changed the tests accordingly.

krlohnes commented:
Thank you for your work on the review @sjudeng!

krlohnes commented:
@davidclement90 This will need a small change to work with TP 3.3.0 `Authenticator` removed deprecated the method `newSaslNegotiator()`, so those will need to be removed from the classes implemented here. Other than that, it's fine.

krlohnes commented:
It also needs `AuthenticationSettings` to be passed in to the constructor of `SaslAndHMACAuthenticator`

sjudeng commented:
@davidclement90 Good question.

@krlohnes I hope you don't mind if we wait for merge of #435 and then ask you to update this for compatibility with TinkerPop 3.3.0.

amcp commented on janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/HMACAuthenticator.java:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    public void setup(final Map<String,Object> config) {
+        if (null != config && !config.containsKey(CONFIG_HMAC_SECRET)) {
```
we have guava in the deps so use Preconditions.checkState()

amcp commented on janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/HMACAuthenticator.java:
```diff
@@ -0,0 +1,222 @@
+        if (timeLong + timeout < currentTime || expected.length() != base64Token.length()) {
+            return false;
+        } else {
+            //Don't short circuit comparison to prevent timing attacks
```
strings are immutable so you could just `return expected.equals(base64Token)`?

krlohnes replied:
We want to check every single character in the individual char array. `expected.equals(base64Token)` leaves this vulnerable to timing attacks e.g. the attacker can figure out which character he's gotten wrong by timing when the unauthorized message comes back.

amcp commented on janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/HMACAuthenticator.java:
```diff
@@ -0,0 +1,222 @@
+      ByteBuffer byteBuffer = Charset.forName("UTF-8").encode(charBuffer);
+      byte[] bytes = Arrays.copyOfRange(byteBuffer.array(),
+              byteBuffer.position(), byteBuffer.limit());
+      Arrays.fill(charBuffer.array(), '\u0000');
```
add a comment as why the fill is necessary?

amcp commented on janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphAbstractAuthenticator.java:
```diff
@@ -0,0 +1,122 @@
+    @Override
+    public void setup(final Map<String,Object> config) {
+        logger.info("Initializing authentication with the {}", this.getClass().getName());
+        if (null == config) {
```
Preconditions.checkArgument()

amcp commented on janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/JanusGraphAbstractAuthenticator.java:
```diff
@@ -0,0 +1,122 @@
+                    this.getClass().getName()));
+        }
+
+        if (!config.containsKey(CONFIG_CREDENTIALS_DB)) {
```
Preconditions.checkState()

amcp commented:
@krlohnes any updates?

sjudeng commented:
The update to TinkerPop 3.3.0 has been merged.

@amcp Could we move the updates to support AuthenticatorProviderChain model and custom authenticators to a separate issue or do you want to see that here? If @krlohnes is able to make those updates I'd think they'd be significant enough we'd still need to dismiss the existing reviews here and re-review at that point.

krlohnes commented:
@amcp I've been away for a while. I agree with @sjudeng with creating a separate PR for the AuthenticatorProviderChain model. It might even make sense to add that to TinkerPop  rather than here in JanusGraph. I am, however, updating with the other changes you've requested and I'm going to fix the 3.3.0 changes.

amcp commented:
Let us move AuthenticatorProviderChain discussion to a separate issue. I will create it.

amcp commented:
@krlohnes the cla tag disappeared. can you rebase and check your commit please?

krlohnes commented:
@amcp I need to submit a new CLA, so that could be correct. I just sent it off to the `janusgraph-cla@googlegroups.com` email. Should I still rebase? Sorry for the delay in response, I'm on a vacation right now.

krlohnes commented:
@mbrukman Do I need to take some action for the new CLA to show up for janusgraph-bot?

mbrukman commented:
@krlohnes — your ICLA has been recorded as part of https://github.com/JanusGraph/legal/pull/79, but you now need to switch the email address used in your commits for it to be recognized as such.

Note that your commits are still using your IBM email address:
https://github.com/JanusGraph/janusgraph/pull/624.patch

To fix this, please follow these instructions:
https://github.com/JanusGraph/janusgraph/blob/master/CONTRIBUTING.md#configure-your-repo-to-match-the-cla
specifically:

1. squash your commits so you have a single commit to deal with, it's easier that way

2. update your git config

    ```bash
    # you've already done `git config user.name` so skip this step
    git config user.email "your-email@gmail.com"
    ```

3. rewrite your (now-single) commit to use the new info

    ```bash
    git commit --amend -s --reset-author   # remove the previous Signed-off-by line
    ```

4. push the updates to GitHub

    ```bash
    git push -f
    ```

Hope this helps.

krlohnes commented:
@mbrukman Thanks. It looks like that worked!

sjudeng commented on docs/basics.adoc:
````diff
@@ -635,22 +635,131 @@ bin/gremlin-server.sh ./conf/gremlin-server/http-gremlin-server.yaml
+```
+IMPORTANT: Note the hmacSecret here. This should be the same across all running JanusGraph servers if you want to be able to use the same HMAC token on each server.
+
+This creates a `/session` endpoint that provides a token that expires after an hour by default. This timeout for the token can be configured through the `tokenTimeout` configuration option in the `authentication.config` map. This value is a Long value and in milliseconds.
````
Does this still work under TinkerPop 3.3? I'm getting the below error trying to get the token.

```
$ curl http://localhost:8182/session -XGET -u user:password
{"message":"no gremlin script supplied"}
```

The test gremlin-server.yaml includes the following.

```channelizer: org.apache.tinkerpop.gremlin.server.channel.HttpChannelizer
graphs: {
  graph: conf/gremlin-server/janusgraph-cassandra-es-server.properties
}
authentication: {
  authenticator: org.janusgraph.graphdb.tinkerpop.gremlin.server.auth.HMACAuthenticator,
  authenticationHandler: org.janusgraph.graphdb.tinkerpop.gremlin.server.handler.HttpHMACAuthenticationHandler,
  config: {
    defaultUsername: user,
    defaultPassword: password,
    hmacSecret: secret,
    credentialsDb: conf/janusgraph-cassandra-es.properties
   }
}
```

krlohnes replied:
@sjudeng This looks like it's a regression issue in TinkerPop 3.3.x I'll fix that there.

sjudeng replied:
Then documentation here should probably be updated. Standalone HTTP HMAC authentication presumably won't be supported in JanusGraph until it's updated to a future TinkerPop release where the issue is addressed. My suggestion is to remove the "HTTP HMAC authentication" section from the documentation for now but move the relevant content to the "Authentication over HTTP and WebSocket" section since it does apply (and work) there. Then we can create a separate issue to track updates to support standalone HTTP HMAC.

mbrukman commented:
@krlohnes — given the length of time this PR has been outstanding (since Oct), could you please update the `master` branch in your clone, rebase your PR on `master`, and `git push -f` again so we get Travis CI to retest your changes with all the intervening updates that have been made in the past few months?

We've had situations in the past where a PR was tested and worked at the time it was proposed, but given a bunch of intervening changes (due to the PR languishing), once it was merged to `master`, ended up breaking the build, so we had to revert it. I'd like to avoid doing that again.

krlohnes commented:
@mbrukman I've rebased on top of master. It should be fine, it's fairly separate work that doesn't touch much else in JanusGraph. @sjudeng I've updated the documentation, removing the standalone HMAC from the documentation and added an issue in JG. I've also added an issue to the TinkerPop Jira and submitted a PR fixing the regression issue.

mbrukman commented:
@krlohnes — this change passed while it was in PR form, but looks like it failed once it was merged on `master`: see [this build](https://travis-ci.org/JanusGraph/janusgraph/builds/324213029), specifically the [Solr 6 job](https://travis-ci.org/JanusGraph/janusgraph/jobs/324213038).

Here's the relevant part of the log:

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.janusgraph.diskstorage.solr.ThriftSolrTest
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/home/travis/.m2/repository/org/slf4j/slf4j-log4j12/1.7.12/slf4j-log4j12-1.7.12.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/home/travis/.m2/repository/ch/qos/logback/logback-classic/1.1.2/logback-classic-1.1.2.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
Tests run: 29, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 407.027 sec - in org.janusgraph.diskstorage.solr.ThriftSolrTest
Running org.janusgraph.diskstorage.solr.BerkeleySolrTest
Tests run: 29, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 267.188 sec - in org.janusgraph.diskstorage.solr.BerkeleySolrTest
Running org.janusgraph.diskstorage.solr.SolrIndexTest
66 vs 73.867869045
Query time on 30000 docs (ms): 50
Tests run: 19, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 28.844 sec <<< FAILURE! - in org.janusgraph.diskstorage.solr.SolrIndexTest
testTTL(org.janusgraph.diskstorage.solr.SolrIndexTest)  Time elapsed: 6.171 sec  <<< FAILURE!
java.lang.AssertionError: expected:<2> but was:<4>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at org.janusgraph.diskstorage.indexing.IndexProviderTest.testTTL(IndexProviderTest.java:704)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
Results :
Failed tests: 
  SolrIndexTest>IndexProviderTest.testTTL:704 expected:<2> but was:<4>
Tests run: 77, Failures: 1, Errors: 0, Skipped: 3
```

Should we rollback this change, or do you have a fix to this you can provide?

sjudeng commented:
@mbrukman Unfortunately the solr jobs in Travis are failing sporadically unrelated to any code changes. You can restart the individual failed job (not the whole Travis build) and it should pass.

krlohnes commented:
@sjudeng  @mbrukman Thanks for that heads up. I was just about to sit down and take a look. I was honestly a little confused since all of the new code in this PR is in the new `janusgraph-server` module.

martingg88 commented:
When do you think this auth feature can be release?
